src/pip/_internal/utils/parallel.py

```diff
@@ -0,0 +1,215 @@
+
+    Return an unordered iterator of the results.
+    """
+    with ProcessPool() as pool, closing(pool):
+        return pool.imap_unordered(func, iterable, chunksize)
+
+
+def _map_multithread_py2(func, iterable, chunksize=None):
+    # type: (Callable[[S], T], Iterable[S], Optional[int]) -> List[T]
+    """Chop iterable into chunks and submit them to a thread pool.
+
+    The (approximate) size of these chunks can be specified
+    by setting chunksize to a positive integer.
+
+    Note that this function may cause high memory usage
+    for long iterables.
+
+    Return a list of results in order.
+    """
+    pool = ThreadPool(DEFAULT_POOLSIZE)
+    try:
+        return pool.map_async(func, iterable, chunksize).get(TIMEOUT)
+    finally:
+        pool.terminate()
+
+
+def _map_multithread_py3(func, iterable, chunksize=None):
+    # type: (Callable[[S], T], Iterable[S], Optional[int]) -> List[T]
+    """Chop iterable into chunks and submit them to a thread pool.
+
+    The (approximate) size of these chunks can be specified
+    by setting chunksize to a positive integer.
+
+    Note that this function may cause high memory usage
+    for long iterables.
+
+    Return a list of results in order.
+    """
+    with ThreadPool(DEFAULT_POOLSIZE) as pool:
+        return pool.map(func, iterable, chunksize)
+
+
+def _imap_multithread_py2(func, iterable, chunksize=1):
+    # type: (Callable[[S], T], Iterable[S], int) -> Iterator[T]
+    """Chop iterable into chunks and submit them to a thread pool.
+
+    For very long iterables using a large value for chunksize can make
+    the job complete much faster than using the default value of 1.
+
+    Return an unordered iterator of the results.
+    """
+    pool = ThreadPool(DEFAULT_POOLSIZE)
+    try:
+        return pool.map_async(func, iterable, chunksize).get(TIMEOUT)
+    finally:
+        pool.terminate()
+
+
+def _imap_multithread_py3(func, iterable, chunksize=1):
+    # type: (Callable[[S], T], Iterable[S], int) -> Iterator[T]
+    """Chop iterable into chunks and submit them to a thread pool.
+
+    For very long iterables using a large value for chunksize can make
+    the job complete much faster than using the default value of 1.
+
+    Return an unordered iterator of the results.
+    """
+    with ThreadPool(DEFAULT_POOLSIZE) as pool, closing(pool):
+        return pool.imap_unordered(func, iterable, chunksize)
+
+
+if LACK_SEM_OPEN:
+    map_multiprocess = map_multithread = _map_fallback
+    imap_multiprocess = imap_multithread = _imap_fallback
+elif PY2:
+    map_multiprocess = _map_multiprocess_py2
+    imap_multiprocess = _imap_multiprocess_py2
+    map_multithread = _map_multithread_py2
+    imap_multithread = _imap_multithread_py2
+else:
+    map_multiprocess = _map_multiprocess_py3
+    imap_multiprocess = _imap_multiprocess_py3
+    map_multithread = _map_multithread_py3
+    imap_multithread = _imap_multithread_py3
```

Comment:
To this end, I'm also 100% OK with just falling back to _imap_fallback on Python 2 (and dropping those implementations) since trying to get things working on Python 2 when we're gonna drop support real soon isn't gonna be worth it.